server/broadcast_test.go

```diff
@@ -519,3 +545,212 @@ func TestVerifyPixels(t *testing.T) {
+		sessManager: bsm,
+	}
+
+	seg := &stream.HLSSegment{}
+	err = transcodeSegment(cxn, seg, "dummy", segmentVerifier)
+	assert.Nil(err)
+	assert.Equal(1, verifier.calls)
+	require.NotNil(verifier.params)
+	assert.Equal(cxn.mid, verifier.params.ManifestID)
+	assert.Equal(seg, verifier.params.Source)
+	// Do it again for good measure
+	err = transcodeSegment(cxn, seg, "dummy", segmentVerifier)
+	assert.Nil(err)
+	assert.Equal(2, verifier.calls)
+
+	// now "disable" the verifier and ensure no calls
+	err = transcodeSegment(cxn, seg, "dummy", nil)
+	assert.Nil(err)
+	assert.Equal(2, verifier.calls)
+
+	// Pass in a nil policy
+	err = transcodeSegment(cxn, seg, "dummy", verification.NewSegmentVerifier(nil))
+	assert.Nil(err)
+
+	// Pass in a policy but no verifier specified
+	policy = &verification.Policy{}
+	err = transcodeSegment(cxn, seg, "dummy", verification.NewSegmentVerifier(policy))
+	assert.Nil(err)
+}
+
+func TestVerifier_Verify(t *testing.T) {
+	assert := assert.New(t)
+
+	cxn := &rtmpConnection{}
+	sess := &BroadcastSession{}
+	source := &stream.HLSSegment{}
+	res := &net.TranscodeData{}
+	verifier := verification.NewSegmentVerifier(&verification.Policy{})
+	URIs := []string{}
+	err := verify(verifier, cxn, sess, source, res, URIs)
+	assert.Nil(err)
+
+	// Test local OS: Should fail with an invalid path
+	sess.BroadcasterOS = drivers.NewMemoryDriver(nil).NewSession("streamName")
+	URIs = append(URIs, "filename")
+	err = verify(verifier, cxn, sess, source, res, URIs)
+	assert.NotNil(err)
+	assert.Contains(err.Error(), "invalid URI for request")
+
+	// Test local OS : Should fail if data does not exist in OS
+	URIs[0] = "/filename"
+	err = verify(verifier, cxn, sess, source, res, URIs)
+	assert.NotNil(err)
+	assert.Contains(err.Error(), "Missing Local Data")
+
+	// Test for segment not in broadcaster's own OS - "livepeer" S3 bucket
+	drivers.S3BUCKET = "livepeer"
+	sess.BroadcasterOS = drivers.NewS3Driver("", drivers.S3BUCKET, "", "").NewSession("")
+	err = verify(verifier, cxn, sess, source, res, URIs)
+	assert.NotNil(err)
+	assert.Contains(err.Error(), "Expected local storage but did not have it")
+
+	// Set broadcaster's OS to "livepeer" S3 bucket and fix the URL
+	URIs[0] = "https://livepeer.s3.amazonaws.com"
+	err = verify(verifier, cxn, sess, source, res, URIs)
+	assert.Nil(err)
+
+	// Check non-retryable errors
+	sess.OrchestratorInfo = &net.OrchestratorInfo{Transcoder: "asdf"}
+	bsm := bsmWithSessList([]*BroadcastSession{sess})
+	cxn.sessManager = bsm
+	sv := &stubVerifier{err: errors.New("NonRetryable")}
+	verifier = newStubSegmentVerifier(sv)
+	assert.Equal(0, sv.calls)  // sanity check initial call count
+	assert.Len(bsm.sessMap, 1) // sanity check initial bsm map
+	err = verify(verifier, cxn, sess, source, res, URIs)
+	assert.NotNil(err)
+	assert.Equal(1, sv.calls)
+	assert.Equal(sv.err, err)
+	assert.Len(bsm.sessMap, 1) // No effect on map for now
+
+	// Check non-retryable errors, esp broadcast session removal from manager
```

Comment:
Should this say retryable errors?

Reply:
Fixed in https://github.com/livepeer/go-livepeer/pull/1096/commits/788ed40861b2fbbc7fcf384d33d77c533abf0345